src/engraving/rendering/dev/accidentalslayout.cpp

```diff
@@ -0,0 +1,1118 @@
+    : allAccidentals(acc), chords(ch)
+{
+    sortTopDown(allAccidentals);
+    stackedAccidentals.reserve(allAccidentals.size());
+    stackedAccidentalsShape.elements().reserve(4 * allAccidentals.size());
+
+    initConstants();
+}
+
+void AccidentalsLayout::AccidentalsLayoutContext::initConstants()
+{
+    m_spatium = allAccidentals.front()->spatium();
+
+    m_verticalAccidentalToAccidentalClearance = 0.15 * m_spatium;
+    m_verticalSharpToSharpClearance = 0.05 * m_spatium;
+    m_verticalAccidentalToChordClearance = 0.10 * m_spatium;
+    m_additionalPaddingForVerticals = 0.1 * m_spatium;
+
+    const MStyle& style = allAccidentals.front()->style();
+    m_accidentalAccidentalDistance = style.styleMM(Sid::accidentalDistance);
+    m_orderFollowNoteDisplacement = style.styleB(Sid::accidentalOrderFollowsNoteDisplacement);
+    m_alignOctavesAcrossSubChords = style.styleB(Sid::alignAccidentalOctavesAcrossSubChords);
+    m_keepSecondsTogether = style.styleB(Sid::keepAccidentalSecondsTogether);
+    m_alignOffsetOctaves = style.styleB(Sid::alignOffsetOctaveAccidentals);
+}
+
+void AccidentalsLayout::layoutAccidentals(const std::vector<Chord*>& chords, LayoutContext& ctx)
+{
+    std::vector<Accidental*> allAccidentals = collectAccidentals(chords);
+
+    if (allAccidentals.empty()) {
+        return;
+    }
+
+    for (Accidental* acc : allAccidentals) {
+        acc->computeMag();
+        TLayout::layoutAccidental(acc, acc->mutldata(), ctx.conf());
+    }
+
+    AccidentalsLayoutContext accidentalsLayoutContext(allAccidentals, chords);
+    doAccidentalPlacement(accidentalsLayoutContext);
+}
+
+std::vector<Accidental*> AccidentalsLayout::collectAccidentals(const std::vector<Chord*> chords)
+{
+    std::vector<Accidental*> accidentals;
+
+    for (Chord* chord : chords) {
+        for (Note* note : chord->notes()) {
+            Accidental* acc = note->accidental();
+            if (!acc) {
+                continue;
+            }
+            acc->setPos(0.0, 0.0);
+            bool isRedundant = false;
+            for (Accidental* otherAcc : accidentals) {
+                if (otherAcc->accidentalType() == acc->accidentalType() && otherAcc->line() == acc->line()) {
+                    isRedundant = true;
+                    break;
+                }
+            }
+            if (isRedundant) {
+                acc->setbbox(RectF());
+                continue;
+            }
+            accidentals.push_back(acc);
+        }
+    }
+
+    return accidentals;
+}
+
+void AccidentalsLayout::doAccidentalPlacement(AccidentalsLayoutContext& ctx)
+{
+    createChordsShape(ctx);
+
+    findOctaves(ctx);
+    findSeconds(ctx);
+    splitIntoSubChords(ctx);
+
+    for (std::vector<Accidental*>& accidentalSubChord : ctx.accidentalSubChords) {
+        layoutSubChord(accidentalSubChord, ctx);
+    }
+
+    verticallyAlignAccidentals(ctx);
+}
+
+void AccidentalsLayout::findOctaves(AccidentalsLayoutContext& ctx)
```

Comment:
Since this code will be called quite often, it is probably better to combine this function with findSeconds. This way, we will reduce the complexity of the algorithm